src/utils/validation.ts

```diff
@@ -1,18 +1,44 @@
 import { FieldProps } from '../form-builder/types';
 import { getMessage } from './i18n';
 
-export const validator = (
-  props: FieldProps<string>
-): ((value: string) => void | undefined | string | Promise<any>) => {
-  return (value: string) => {
-    if (props.required && !value) {
-      const errorMessage =
-        typeof props.required === 'string'
-          ? props.required
-          : getMessage('required.default');
-      return errorMessage;
-    }
+export type ValidationFunctionResult<T> =
+  | void
+  | undefined
+  | string
+  | Promise<any>;
+export type ValidationFunction<T> = (
+  value: T,
+  props: FieldProps<T>
+) => ValidationFunctionResult<T>;
 
+export const chainValidations = <T>(
+  props: FieldProps<T>,
+  validations: ValidationFunction<T>[]
+): ((value: T) => ValidationFunctionResult<T>) => {
+  return (value: T) => {
+    // Return the error message from the first validation function that fails.
+    const errorMessage = validations
+      .map((v) => v(value, props))
+      .filter((m) => m)[0];
+    if (errorMessage) return errorMessage;
```

Comment:
I feel like this behavior needs some kind of documented decision. For example, imagine a user creating a password. They receive a first error message: "Your password must have at least 18 characters". So, they add some more characters to meet the limit, then they receive a _new_ error message: "Your password must contain at least one integer". So the user makes another changes and attempts to confirm, only to be greeted with another message: "Your password must have one of these special characters: ..."

In this example I think it would be ideal to run all validations and concatenate the messages to display:

```
- Your password must have at least 18 characters
- Your password must contain at least one integer
- Your password must have one of these special characters: ...
```

In theory there might be validations which might not initially fail, but I think it may be a better experience if we show them everything that is wrong with the current value so that the user has an attempt to fix it all at once.

Reply:
That behavior is a deviation from our current UX (showing only one message at a time). To be clear, though, the behavior this function provides is such:
1. User enters a value
2. User blurs the field
    - This triggers showing the error message for the first time
    - All subsequent error messages are shown immediately using `Formik`'s default `validateOnChange={true}`, but can be changed
3. Error message is shown
4. User changes the value
5. Second error message is shown (without blurring or submitting)
6. User changes the value
7. Third error message is shown (again, while the field still has focus)